src/System.Management.Automation/engine/Subsystem/PredictionSubsystem/ICommandPredictor.cs

```diff
@@ -131,6 +131,12 @@ public sealed class PredictionClient
         /// </summary>
         public PredictionClientKind Kind { get; }
 
+        /// <summary>
+        /// Gets the current location of the default session.
+        /// It returns null if there is no default Runspace or if the default is a remote Runspace.
+        /// </summary>
+        public PathInfo? CurrentLocation { get; set; }
```

Comment:
I thought about having this information in `PredictionContext`, but it's difficult to do so for 2 folds of reasons:
1. The `PredictionContext` is constructed within `CommandPrediction.PredictInputAsync`, which is a static API. So, to get the current location, you will have to use `Runspace.DefaultRunspace`, which is not guaranteed to exist on the calling thread. Also, it makes it hard to check if the host is currently in a nested prompt talking to a remote session.

2. Having the current location information in `PredictionContext` means we need to retrieve that information every time when `CommandPrediction.PredictInputAsync` is invoked, namely for every keystroke when the user is editing the current command line. That will be a lot of unnecessary work because the current location won't change after the PowerShell host calls into PSReadLine, so it's always the same while user is still editing the current line.

The information about both the current location and whether the host is talking to a remote Runspace is very easy to get in PSReadLine itself, and we only need to update the information once per every call to `PSReadLine.ReadLine` and use that for all the `PredictInputAsync` calls before returning from `PSReadLine.ReadLine`.

`PredictionClient` essentially is information about the host and the information the host can provides. Even though the current location is an engine state, the host can easily access it and has the knowledge to decide whether to provide this information (_don't set the current location when a remote Runspace is pushed, as the local location doesn't help for command lines that are going to run in the remote Runspace_). That's why I decided to have the `CurrentLocation` property in `PredictionClient`.

Reply:
Ah yep that's the missing context, makes sense to me. Thanks for the explanation, I'll mark this resolved ❤️